kep/42-podgroup-coscheduling/README.md

```diff
@@ -135,12 +141,75 @@ cluster upgrade.
 
 #### QueueSort
 
-To make sure a group of pods can be scheduled as soon as possible. We implemented this `extension point`. The main progress is as follows:
+For Pods belong to a same podGroup should be popped out of the scheduling queue one by one,
+which requires an efficient and wise algorithm. This is quite different with plain Pods,
+so we need to reimplement the QueueSort plugin.
 
-1. Sort based on Pod Priority
-2. Pod does not belong to any group
-3. PodGroup creation time
-4. Pod creation time
+Basically, we need to watch for all the podGroups and maintain a cache for them
+in coscheduling's core package:
+
+- when podGroup created, we'll create a `queuedPodGroup` in the cache
+- when podGroup updated, we'll do nothing right now
+- when podGroup deleted, we'll clean up the objet in the cache
+
+The structure looks like below:
+
+```golang
+type PodGroupManager struct {
+ // ...
+ // <new>
+ // key is the podGroup name, value is the queued podGroup info.
+ podGroups map[string]*queuedPodGroup
+}
+
+// queuedPodGroup's lifecycle is synchronized with the podGroup object.
+type queuedPodGroup struct {
+ // Timestamp is the podGroup's queued time.
+ // - timestamp will be initialized when the first pod enqueues
+ // - timestamp will be renewed when a pod re-enqueues for failing the scheduling
+ timestamp time.Time
+
+ // status indicated the podGroup's condition, it can be:
+ // - queueing: when the first pod starts scheduling
+ // - queueingFailed: when a pod fails the scheduling
+ // - queueingSucceeded: when the podGroup succeeds in scheduling
+ status string
+}
+```
+
+The workflow is (only works for pods have podGroup):
+
+1. When a new podGroup created, we'll create a queuedPodGroup in the cache.
+2. When the first pod of the podGroup creates, in queueing specifically, we'll initialize the timestamp of the queuedPodGroup,
+also set the status=queueing. This will happen in `Less()` function, and because we'll not change the timestamp, so pods belong
+to the same podGroup will be placed together as a unit.
+3. Schedule the pods of the podGroup one by one, if anyone fails in the scheduling cycle, we'll set the status=queueingFailed in PostFilter,
+and the podGroup will enter into the backoff queue if configured.
+4. When the failed Pod re-enqueues, we'll renew the queuedPodGroup timestamp based on the `queueingFailed` status and also set the status to `queueing`, the new timestamp should be `now+backoffTime` to avoid breaking the podGroup who starts scheduling
```

Comment:
do you mean to renew during QueueSort()? If so, we shouldn't do that. See a comment below.

I'd suggest is to renew the timestamp at the end when a group of pods failed as a unit; instead of the beginning when the group starts its next scheduling attempt.